adr-11-centralized-sequencer.md

```diff
@@ -0,0 +1,213 @@
+
+A decentralized sequencer would distribute the sequencing responsibility across multiple nodes, providing better censorship resistance and fault tolerance. However, this approach introduces significant complexity in terms of consensus, leader election, and coordination between nodes. It would also require more resources to operate and maintain.
+
+This approach was not chosen for the initial implementation because:
+1. It adds unnecessary complexity for many use cases
+2. It requires more development time and resources
+3. Many rollup projects start with a centralized sequencer and gradually move towards decentralization
+
+### Embedded Sequencer in Rollup Nodes
+
+Another approach would be to embed sequencing functionality directly into rollup nodes. This would simplify the architecture by eliminating a separate sequencer component.
+
+This approach was not chosen because:
+1. It couples sequencing logic with rollup node logic, reducing modularity
+2. It makes it harder to upgrade or replace the sequencing component independently
+3. It doesn't allow for a dedicated sequencing service that can be optimized separately
+
+## Decision
+
+We implement a standalone centralized sequencer that:
+
+1. Implements the Generic Sequencer interface from the `go-sequencing` package
+2. Exposes a gRPC service for rollup clients to submit transactions
+3. Batches transactions and submits them to a DA layer at regular intervals
+4. Maintains state to track submitted batches and provide verification
+5. Provides metrics for monitoring and observability
+
+The centralized sequencer is a separate repository and can be deployed as a standalone service or as a Docker container.
+
+## Detailed Design
+
+### User Requirements
+
+- Rollup developers need a simple, reliable sequencer that can order transactions and submit them to a DA layer
+- The sequencer should be easy to deploy and configure
+- The sequencer should provide metrics for monitoring
+- The sequencer should be able to recover from crashes and maintain state
+
+### Systems Affected
+
+- Rollup nodes that interact with the sequencer
+- DA layer where batches are submitted
+
+### Data Structures
+
+The centralized sequencer uses the following key data structures:
+
+1. **BatchQueue**: A queue to store batches of transactions waiting to be processed
+   ```go
+   type BatchQueue struct {
+       queue []sequencing.Batch
+       mu    sync.Mutex
+   }
+   ```
+
+2. **Sequencer**: The main sequencer structure that implements the Generic Sequencer interface
+   ```go:centralized-sequencer/adr-11-centralized-sequencer.md
+   type Sequencer struct {
+       da              goda.DA
+       namespace       []byte
+       rollupId        []byte
+       batchTime       time.Duration
+       bq              *BatchQueue
+       lastBatchHash   []byte
+       seenBatches     map[string]struct{}
+       seenBatchesMutex sync.RWMutex
+       metrics         *Metrics
+       db              *badger.DB
```

Comment:
do we need this to have storage?

Reply:
The only reason to have storage currently is to keep track of the last batch hash, which is needed when a rollkit node is restarted. We can store this in Rollkit's store as well.